hydroflow_lang/src/graph/ops/source_interval.rs

```diff
@@ -0,0 +1,86 @@
+use quote::quote_spanned;
+use syn::parse_quote_spanned;
+
+use crate::graph::OperatorInstance;
+
+use super::{
+    FlowProperties, FlowPropertyVal, OperatorConstraints, OperatorWriteOutput, WriteContextArgs,
+    RANGE_0, RANGE_1,
+};
+
+/// > 0 input streams, 1 output stream
+///
+/// > Arguments: A [`Duration`](https://doc.rust-lang.org/stable/std/time/struct.Duration.html) for this interval.
+///
+/// Emits [Tokio time `Instant`s](https://docs.rs/tokio/1/tokio/time/struct.Instant.html) on a
+/// repeated interval. The first tick completes imediately. Missed ticks will be scheduled as soon
+/// as possible, and the `Instant` will be the missed time, not the late time.
+///
+/// Note that this requires the hydroflow instance be run within a [Tokio `Runtime`](https://docs.rs/tokio/1/tokio/runtime/struct.Runtime.html).
+/// The easiest way to do this is with a [`#[tokio::main]`](https://docs.rs/tokio/1/tokio/attr.main.html)
+/// annotation on `async fn main() { ... }`.
+///
+/// ```rustbook
+/// use std::time::Duration;
+///
+/// use hydroflow::hydroflow_syntax;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut hf = hydroflow_syntax! {
+///         source_interval(Duration::from_secs(1))
+///             -> for_each(|time| println!("This runs every second: {:?}!", time));
+///     };
+///
+///     // Will print 4 times (fencepost counting).
+///     tokio::time::timeout(Duration::from_secs_f32(3.5), hf.run_async())
+///         .await
+///         .expect_err("Expected time out");
+/// }
+/// ```
+#[hydroflow_internalmacro::operator_docgen]
+pub const SOURCE_INTERVAL: OperatorConstraints = OperatorConstraints {
+    name: "source_interval",
+    hard_range_inn: RANGE_0,
+    soft_range_inn: RANGE_0,
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 1,
+    persistence_args: RANGE_0,
+    type_args: RANGE_0,
+    is_external_input: true,
+    ports_inn: None,
+    ports_out: None,
+    properties: FlowProperties {
+        deterministic: FlowPropertyVal::Yes,
+        monotonic: FlowPropertyVal::Yes,
+        inconsistency_tainted: false,
+    },
+    input_delaytype_fn: |_| None,
+    write_fn: |wc @ &WriteContextArgs {
+                   root,
+                   op_span,
+                   op_inst: OperatorInstance { arguments, .. },
+                   ..
+               },
+               diagnostics| {
+        let ident_intervalstream = wc.make_ident("intervalstream");
+        let mut write_prologue = quote_spanned! {op_span=>
+            let #ident_intervalstream =
+                #root::tokio_stream::wrappers::IntervalStream::new(#root::tokio::time::interval(#arguments));
```

Comment:
I guess a type check here doesn't improve the error message materially? Like if they put in an integer instead of a duration?

Reply:
That's actually a good question, I'll add error message tests and add a check if needed

Reply:
Yeah the error message looks good already:
```rust
error[E0308]: mismatched types
 --> tests/compile-fail/surface_source_interval_badarg.rs:7:25
  |
5 |       let mut df = hydroflow_syntax! {
  |  __________________-
6 | |         // Should be a `Duration`.
7 | |         source_interval(5) -> for_each(std::mem::drop);
  | |                         ^ expected `Duration`, found integer
8 | |     };
  | |_____- arguments to this function are incorrect
  |
note: function defined here
 --> $CARGO/tokio-1.26.0/src/time/interval.rs
  |
  | pub fn interval(period: Duration) -> Interval {
  |        ^^^^^^^^
  ```